src/qibo/hamiltonians/hamiltonians.py

```diff
@@ -564,7 +564,9 @@ def expectation_from_samples(self, freq, qubit_map=None):
             if len(term.factors) != len(set(term.factors)):
                 raise_error(NotImplementedError, "Z^k is not implemented since Z^2=I.")
         keys = list(freq.keys())
-        counts = np.array(list(freq.values())) / sum(freq.values())
+        counts = self.backend.cast(list(freq.values()), self.backend.precision) / sum(
```

Comment:
Why the pass `dtype=backend.precision`? Wouldn't it be `dtype=self.backend.dtype`?

Reply:
No because `backend.dtype` specifies the type of the states and matrices, which is going to be `complex128/64`, whereas the `backend.precision` only specifies `"single"` or `"double"` which casts to `float32/64` that is what we need for expectation values.